server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -658,17 +879,49 @@ private long appendAddColumns(final View view, final long startRange, final long
-                drainableColumn.visitBuffers(bufferListener);
-
-                // Add the drainable last as it is allowed to immediately close a row set the visitors need
-                addStream.accept(drainableColumn);
+        try (final WritableRowSet myAddedOffsets = view.addRowOffsets().subSetByPositionRange(startRange, endRange)) {
+            // every column must write to the stream
+            for (final ChunkListInputStreamGenerator chunkSetGen : addColumnData) {
+                boolean columnWritten = false;
+
+                // iterate through each chunk and find the one that contains the requested data
+                for (int i = 0; i < chunkSetGen.size() && !columnWritten; ++i) {
+                    // if the set of data is empty or the requested rows start in this chunk
+                    if (startRange <= addChunkRowSets[i].lastRowKey()) {
+                        final ChunkInputStreamGenerator generator = chunkSetGen.generators[i];
+
+                        // shift this into the chunk position space
+                        final long shiftAmount = -addChunkRowSets[i].firstRowKey();
+
+                        // get an offset RowSet for each row in this chunk
+                        try (final WritableRowSet adjustedOffsets = myAddedOffsets.intersect(addChunkRowSets[i])) {
+                            // normalize this to the chunk offsets
+                            adjustedOffsets.shiftInPlace(shiftAmount);
+
+                            final ChunkInputStreamGenerator.DrainableColumn drainableColumn =
+                                    generator.getInputStream(view.options(), adjustedOffsets);
+                            drainableColumn.visitFieldNodes(fieldNodeListener);
+                            drainableColumn.visitBuffers(bufferListener);
+
+                            // Add the drainable last as it is allowed to immediately close a row set the visitors need
+                            addStream.accept(drainableColumn);
+                        }
+                        // no need to test other chunks
+                        columnWritten = true;
+                    }
+                }
+                // must write column data, just write an empty column
+                if (!columnWritten) {
+                    try (final RowSet empty = RowSetFactory.empty()) {
+                        final ChunkInputStreamGenerator.DrainableColumn drainableColumn =
+                                chunkSetGen.generators[0].getInputStream(view.options(), empty);
+                        drainableColumn.visitFieldNodes(fieldNodeListener);
+                        drainableColumn.visitBuffers(bufferListener);
+
+                        // Add the drainable last as it is allowed to immediately close a row set the visitors need
+                        addStream.accept(drainableColumn);
+                    }
+                }
             }
             return myAddedOffsets.size();
```

Comment:
The return value needs to be the number of rows written in this record batch, which is `adjustedOffsets.size()`.